File: pratap_crm_ai/pratap_crm_ai/page/crm_ai/crm_ai.py
# Copyright (c) 2026, Viv Choudhary and contributors
# For license information, please see license.txt

import base64
import json
import os
import re

import frappe
import requests

SETTINGS_DOCTYPE = "Pratap CRM Ai Settings"
DEFAULT_OPENAI_MODEL = "gpt-4.1-nano"

PROSPECT_AI_FIELDS = (
	"company_name",
	"custom_contact_person",
	"custom_mobile_no",
	"custom_email",
	"custom_city",
	"custom_postal_code",
	"custom_country",
	"custom_remarks",
)

EXTRACT_PROMPT = (
	"Extract all details from the business card image(s). "
	"Return pure JSON only with these exact keys: "
	"company_name, custom_contact_person, custom_mobile_no, custom_email, "
	"custom_city, custom_postal_code, custom_country, custom_remarks. "
	"Use company_name for the organization on the card. "
	"Use custom_contact_person for the person's full name. "
	"Use custom_remarks for notes, remarks, or other text not mapped above. "
	"Put any remaining extra text under additional_info."
)


@frappe.whitelist()
def get_ai_settings():
	"""Return safe OpenAI settings for the CRM AI page (no API key)."""
	settings = frappe.get_single(SETTINGS_DOCTYPE)
	api_key = (settings.get("openai_api_key") or "").strip()
	model = (settings.get("openai_model") or "").strip() or DEFAULT_OPENAI_MODEL

	return {
		"enable_openai": bool(settings.get("enable_openai")),
		"openai_configured": bool(api_key and model),
		"openai_model": model,
	}


def _get_openai_config():
	settings = frappe.get_single(SETTINGS_DOCTYPE)
	if not settings.get("enable_openai"):
		frappe.throw(
			"OpenAI is disabled. Enable it in Pratap CRM Ai Settings.",
			title="AI Disabled",
		)

	api_key = (settings.get("openai_api_key") or "").strip()
	model = (settings.get("openai_model") or "").strip() or DEFAULT_OPENAI_MODEL

	if not api_key:
		frappe.throw("OpenAI API key is not configured in Pratap CRM Ai Settings.")
	if not model:
		frappe.throw("OpenAI model is not set in Pratap CRM Ai Settings.")

	return api_key, model


def _image_path_to_data_url(image_ref):
	if not image_ref:
		return None

	if image_ref.startswith("data:"):
		return image_ref

	if image_ref.startswith("http"):
		return requests.utils.requote_uri(image_ref)

	if not image_ref.startswith("/"):
		return frappe.utils.get_url(image_ref)

	is_private = "/private/files/" in image_ref
	filename = image_ref.split("/files/")[-1]
	file_path = frappe.get_site_path(
		"private" if is_private else "public", "files", filename
	)

	if not os.path.exists(file_path):
		return frappe.utils.get_url(image_ref)

	try:
		from io import BytesIO

		from PIL import Image

		with Image.open(file_path) as img:
			if img.mode in ("RGBA", "P"):
				img = img.convert("RGB")
			img.thumbnail((1024, 1024))
			buffer = BytesIO()
			img.save(buffer, format="JPEG", quality=80)
			encoded = base64.b64encode(buffer.getvalue()).decode("utf-8")
		return f"data:image/jpeg;base64,{encoded}"
	except Exception:
		with open(file_path, "rb") as handle:
			encoded = base64.b64encode(handle.read()).decode("utf-8")
		ext = os.path.splitext(filename)[1].lower().lstrip(".") or "jpeg"
		mime = "png" if ext == "png" else "jpeg"
		return f"data:image/{mime};base64,{encoded}"


def _parse_json_from_text(text):
	if not text:
		return {}
	match = re.search(r"\{[\s\S]*\}", text)
	if not match:
		return {}
	try:
		return json.loads(match.group(0))
	except json.JSONDecodeError:
		return {}


def _normalize_prospect_data(parsed):
	"""Map generic OCR keys to Prospect custom fields."""
	data = dict(parsed or {})

	if not data.get("company_name"):
		data["company_name"] = data.get("company") or data.get("organization")

	if not data.get("custom_contact_person"):
		first = (data.get("first_name") or "").strip()
		last = (data.get("last_name") or "").strip()
		full = " ".join(part for part in (first, last) if part).strip()
		data["custom_contact_person"] = full or data.get("contact_person") or data.get("name")

	if not data.get("custom_mobile_no"):
		data["custom_mobile_no"] = (
			data.get("mobile")
			or data.get("mobile_no")
			or data.get("phone")
			or data.get("whatsapp")
		)

	if not data.get("custom_email"):
		data["custom_email"] = data.get("email") or data.get("email_id")

	if not data.get("custom_city"):
		data["custom_city"] = data.get("city")

	if not data.get("custom_postal_code"):
		data["custom_postal_code"] = (
			data.get("postal_code") or data.get("pincode") or data.get("zip")
		)

	if not data.get("custom_country"):
		data["custom_country"] = data.get("country")

	if isinstance(data.get("custom_remarks"), dict):
		data["custom_remarks"] = json.dumps(data["custom_remarks"], ensure_ascii=False)

	if not data.get("custom_remarks"):
		data["custom_remarks"] = (
			data.get("remarks") or data.get("remark") or data.get("notes")
		)
		if not data.get("custom_remarks") and data.get("additional_info"):
			extra = data["additional_info"]
			data["custom_remarks"] = (
				json.dumps(extra, ensure_ascii=False)
				if isinstance(extra, dict)
				else str(extra)
			)

	return {key: data.get(key) for key in PROSPECT_AI_FIELDS if data.get(key)}


@frappe.whitelist()
def extract_business_card_data(front_image=None, back_image=None):
	"""Extract business card details using OpenAI vision."""
	front_url = _image_path_to_data_url(front_image)
	back_url = _image_path_to_data_url(back_image)

	if not front_url and not back_url:
		frappe.throw("Please upload at least one visiting card image (front or back).")

	api_key, model = _get_openai_config()

	content = [{"type": "input_text", "text": EXTRACT_PROMPT}]
	if front_url:
		content.append({"type": "input_image", "image_url": front_url})
	if back_url:
		content.append({"type": "input_image", "image_url": back_url})

	payload = {
		"model": model,
		"input": [{"role": "user", "content": content}],
	}

	try:
		response = requests.post(
			"https://api.openai.com/v1/responses",
			headers={
				"Content-Type": "application/json",
				"Authorization": f"Bearer {api_key}",
			},
			json=payload,
			timeout=120,
		)
		data = response.json()

		if not response.ok:
			message = data.get("error", {}).get("message", "OpenAI request failed")
			frappe.throw(message)

		text = ""
		for item in data.get("output", []):
			for block in item.get("content", []):
				if block.get("type") == "output_text":
					text += block.get("text", "")

		parsed = _normalize_prospect_data(_parse_json_from_text(text))

		return {
			"status": "success",
			"source": "openai",
			"raw_text": text,
			"parsed_data": parsed,
		}
	except frappe.ValidationError:
		raise
	except Exception as exc:
		frappe.log_error("Pratap CRM AI — OpenAI extract", frappe.get_traceback())
		frappe.throw(f"Error during OpenAI extraction: {exc}")


@frappe.whitelist()
def create_prospect(prospect_data, front_image=None, back_image=None):
	"""Insert parsed data into ERPNext Prospect."""
	if isinstance(prospect_data, str):
		try:
			prospect_data = json.loads(prospect_data)
		except Exception:
			frappe.throw("Invalid JSON for prospect_data")

	company_name = (prospect_data.get("company_name") or "").strip()
	if not company_name:
		frappe.throw(
			_("Company Name is required to create a Prospect."),
			title=_("Missing Company Name"),
		)

	meta = frappe.get_meta("Prospect")
	prospect_fields = {"doctype": "Prospect", "company_name": company_name}

	for fieldname in PROSPECT_AI_FIELDS:
		if fieldname == "company_name" or not meta.has_field(fieldname):
			continue
		value = (prospect_data.get(fieldname) or "").strip()
		if value:
			prospect_fields[fieldname] = value

	if meta.has_field("company") and not prospect_fields.get("company"):
		prospect_fields["company"] = frappe.defaults.get_user_default("Company")

	doc = frappe.get_doc(prospect_fields)
	doc.company_name = company_name
	doc.insert(ignore_permissions=True, ignore_mandatory=True)

	if not frappe.db.get_value("Prospect", doc.name, "company_name"):
		frappe.db.set_value("Prospect", doc.name, "company_name", company_name, update_modified=False)

	if front_image:
		file_url = _save_image(front_image, f"{doc.name}_front")
		if file_url:
			_attach_file_to_prospect(doc.name, file_url)

	if back_image:
		file_url = _save_image(back_image, f"{doc.name}_back")
		if file_url:
			_attach_file_to_prospect(doc.name, file_url)

	frappe.db.commit()
	return doc.name


def _save_image(image_data, file_base):
	if image_data.startswith("/") and "/files/" in image_data:
		return image_data

	if image_data.startswith("data:image"):
		try:
			header, encoded = image_data.split(",", 1)
			file_ext = "png" if "png" in header else "jpg"
			file_name = f"{file_base}.{file_ext}"
			file_path = frappe.utils.get_site_path("public", "files", file_name)
			with open(file_path, "wb") as handle:
				handle.write(base64.b64decode(encoded))
			return f"/files/{file_name}"
		except Exception as exc:
			frappe.log_error(f"Failed to save image {file_base}", str(exc))
			return None

	return None


def _attach_file_to_prospect(prospect_name, file_url):
	frappe.get_doc(
		{
			"doctype": "File",
			"file_url": file_url,
			"attached_to_doctype": "Prospect",
			"attached_to_name": prospect_name,
			"is_private": 0,
			"file_name": os.path.basename(file_url),
		}
	).insert(ignore_permissions=True)

File: pratap_crm_ai/pratap_crm_ai/page/crm_ai/crm_ai.js
frappe.pages["crm-ai"].on_page_load = function (wrapper) {
	const PAGE_METHOD = "pratap_crm_ai.pratap_crm_ai.page.crm_ai.crm_ai";

	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Business Card Scanner — Prospect"),
		single_column: true,
	});

	page.set_primary_action(__("View Prospects"), () => frappe.set_route("List", "Prospect"), "list");
	page.set_secondary_action(__("New scan"), () => resetForm(true), "refresh");

	$(`
		<div class="pratap-crm-ai-page row p-3">
			<div class="col-lg-6">
				<div class="card p-3 shadow-sm mb-3">
					<h5 class="mb-3">${__("Visiting card images")}</h5>
					<div id="card_image_fields"></div>
					<button id="extractBtn" class="btn btn-dark w-100 mt-3">
						<span class="mr-1">🔍</span> ${__("Extract & Auto-Fill")}
					</button>
					<div class="row mt-3">
						<div class="col-6">
							<div class="small text-muted fw-semibold mb-2">${__("Front preview")}</div>
							<div id="front_preview" class="text-center border rounded bg-light p-2" style="min-height:6rem;">
								<span class="text-muted small">${__("No image yet")}</span>
							</div>
						</div>
						<div class="col-6">
							<div class="small text-muted fw-semibold mb-2">${__("Back preview")}</div>
							<div id="back_preview" class="text-center border rounded bg-light p-2" style="min-height:6rem;">
								<span class="text-muted small">${__("No image yet")}</span>
							</div>
						</div>
					</div>
					<div id="ocr-status" class="small text-muted mt-2"></div>
					<div id="ai-settings-banner" class="alert alert-warning py-2 px-3 small mt-2" style="display:none;"></div>
				</div>
			</div>
			<div class="col-lg-6">
				<div class="card p-3 shadow-sm mb-3" id="prospect-output">
					<div id="prospect-success-slot" class="mb-0"></div>
					<h4 class="mb-3">📇 ${__("Prospect Details")}</h4>
					<div id="prospect-fields" class="mb-3"></div>
					<div id="extra-info" class="mb-3" style="display:none;">
						<h6>${__("Additional info from card")}</h6>
						<div id="extra-info-text" class="bg-light border p-2 rounded small"
							style="max-height:180px;overflow:auto;white-space:pre-wrap;font-family:monospace;"></div>
					</div>
					<button class="btn btn-success w-100" id="saveProspectBtn">
						💾 ${__("Insert into ERPNext Prospect")}
					</button>
					<div id="prospect-insert-msg" class="mt-3"></div>
				</div>
				<div class="card p-3 shadow-sm" id="ocr-raw-card" style="display:none;">
					<h6 class="d-flex justify-content-between align-items-center mb-2">
						${__("Full AI output")}
						<button class="btn btn-sm btn-outline-secondary" id="toggle-ocr">${__("Hide")}</button>
					</h6>
					<pre id="ocr-raw-text" class="bg-light p-2 rounded small mb-0"
						style="max-height:220px;overflow:auto;white-space:pre-wrap;"></pre>
				</div>
			</div>
		</div>
	`).appendTo(page.body);

	const PROSPECT_LAYOUT_FIELDS = [
		{
			fieldname: "company_name",
			fieldtype: "Data",
			label: `🏢 ${__("Company Name")}`,
			reqd: 1,
		},
		{
			fieldname: "custom_contact_person",
			fieldtype: "Data",
			label: `👤 ${__("Contact Person")}`,
		},
		{
			fieldname: "custom_mobile_no",
			fieldtype: "Data",
			label: `📱 ${__("Mobile No")}`,
			options: "Phone",
		},
		{
			fieldname: "custom_email",
			fieldtype: "Data",
			label: `✉️ ${__("Email")}`,
			options: "Email",
		},
		{ fieldname: "__col_break", fieldtype: "Column Break" },
		{
			fieldname: "custom_city",
			fieldtype: "Data",
			label: `📍 ${__("City")}`,
		},
		{
			fieldname: "custom_postal_code",
			fieldtype: "Data",
			label: `📮 ${__("Postal Code")}`,
		},
		{
			fieldname: "custom_country",
			fieldtype: "Data",
			label: `🌍 ${__("Country")}`,
		},
		{
			fieldname: "custom_remarks",
			fieldtype: "Small Text",
			label: `📝 ${__("Remarks")}`,
		},
	];

	const LAYOUT_ONLY = ["Column Break", "Section Break", "Tab Break", "Page Break", "Fold"];
	const PROSPECT_FIELD_KEYS = PROSPECT_LAYOUT_FIELDS.filter(
		(f) => f.fieldname && !LAYOUT_ONLY.includes(f.fieldtype)
	).map((f) => f.fieldname);

	const prospectLayout = new frappe.ui.form.Layout({
		body: page.body.find("#prospect-fields"),
		fields: PROSPECT_LAYOUT_FIELDS,
		doctype: "Prospect",
	});
	prospectLayout.make();
	prospectLayout.refresh();
	const prospectControls = prospectLayout.fields_dict;

	const cardImageLayout = new frappe.ui.form.Layout({
		body: page.body.find("#card_image_fields"),
		fields: [
			{
				fieldname: "front_image",
				fieldtype: "Attach Image",
				label: __("Front of card"),
				options: { make_attachments_public: true, allow_toggle_private: false },
			},
			{ fieldname: "__card_image_col", fieldtype: "Column Break" },
			{
				fieldname: "back_image",
				fieldtype: "Attach Image",
				label: __("Back of card"),
				options: { make_attachments_public: true, allow_toggle_private: false },
			},
		],
		doctype: "Prospect",
	});
	cardImageLayout.make();
	cardImageLayout.refresh();

	const frontCtrl = cardImageLayout.fields_dict.front_image;
	const backCtrl = cardImageLayout.fields_dict.back_image;

	function updatePreview(ctrl, selector) {
		const $box = page.body.find(selector);
		const url = ctrl.get_value();
		if (!url) {
			$box.html(`<span class="text-muted small">${__("No image yet")}</span>`);
			return;
		}
		const fullUrl = frappe.urllib.get_full_url(url);
		const safe = frappe.utils.escape_html(fullUrl);
		$box.html(
			`<a href="${safe}" target="_blank" rel="noopener noreferrer">` +
				`<img src="${safe}" class="img-fluid rounded" style="max-height:160px;object-fit:contain" alt="">` +
				`</a>`
		);
	}

	function watchPreview(parentSel, ctrl, previewSel) {
		const parent = page.body.find(parentSel)[0];
		if (!parent) return;
		new MutationObserver(() => updatePreview(ctrl, previewSel)).observe(parent, {
			subtree: true,
			childList: true,
			attributes: true,
		});
		updatePreview(ctrl, previewSel);
	}

	watchPreview("#card_image_fields", frontCtrl, "#front_preview");
	watchPreview("#card_image_fields", backCtrl, "#back_preview");

	let aiSettings = { enable_openai: false, openai_configured: false };

	function refreshSettingsBanner() {
		const $banner = page.body.find("#ai-settings-banner");
		if (aiSettings.enable_openai && aiSettings.openai_configured) {
			$banner.hide();
			return;
		}
		let msg = __("OpenAI is disabled.");
		if (aiSettings.enable_openai && !aiSettings.openai_configured) {
			msg = __("OpenAI is enabled but API key or model is missing in Pratap CRM Ai Settings.");
		}
		$banner
			.html(`${msg} <a href="/app/pratap-crm-ai-settings">${__("Open settings")}</a>`)
			.show();
	}

	frappe.call({
		method: `${PAGE_METHOD}.get_ai_settings`,
		callback(r) {
			aiSettings = r.message || aiSettings;
			refreshSettingsBanner();
		},
	});

	function showProspectSuccessBanner(prospectName) {
		const safe = frappe.utils.escape_html(prospectName);
		const href = `/app/prospect/${encodeURIComponent(prospectName)}`;
		page.body.find("#prospect-success-slot").html(
			`<div class="alert alert-success prospect-success-banner d-flex justify-content-between align-items-center shadow-sm fade show mb-3"
				style="border-radius:10px;">
				<div>✅ <strong>${__("Prospect")}</strong> <b>${safe}</b> ${__("created successfully.")}</div>
				<a href="${href}" target="_blank" class="btn btn-sm btn-dark">🔗 ${__("Open")}</a>
			</div>`
		);
	}

	function resetForm(showToast = false, keepImages = false, clearSuccessBanner = true) {
		PROSPECT_FIELD_KEYS.forEach((key) => {
			const c = prospectControls[key];
			if (c && c.set_value) c.set_value("");
		});
		page.body.find("#ocr-status, #prospect-insert-msg, #extra-info-text").empty();
		page.body.find("#ocr-raw-text").empty();
		page.body.find("#extra-info, #ocr-raw-card").hide();
		if (clearSuccessBanner) {
			page.body.find("#prospect-success-slot").empty();
		}

		if (!keepImages) {
			frontCtrl.set_value("");
			backCtrl.set_value("");
		}

		prospectLayout.refresh();
		cardImageLayout.refresh();
		updatePreview(frontCtrl, "#front_preview");
		updatePreview(backCtrl, "#back_preview");

		if (showToast) {
			frappe.show_alert({
				message: __("Ready for a new business card scan"),
				indicator: "blue",
			});
		}
	}

	function fillProspectFields(parsed) {
		if (!parsed || typeof parsed !== "object") return;

		let filled = 0;
		for (const key of PROSPECT_FIELD_KEYS) {
			const val = parsed[key];
			if (val && prospectControls[key]) {
				prospectControls[key].set_value(val);
				filled++;
			}
		}

		if (
			parsed.additional_info &&
			!prospectControls.custom_remarks?.get_value?.()
		) {
			const extra =
				typeof parsed.additional_info === "string"
					? parsed.additional_info
					: JSON.stringify(parsed.additional_info, null, 2);
			if (prospectControls.custom_remarks) {
				prospectControls.custom_remarks.set_value(extra);
				filled++;
			}
		}

		if (parsed.additional_info && Object.keys(parsed.additional_info).length) {
			page.body.find("#extra-info").show();
			page.body.find("#extra-info-text").text(
				typeof parsed.additional_info === "string"
					? parsed.additional_info
					: JSON.stringify(parsed.additional_info, null, 2)
			);
		} else {
			page.body.find("#extra-info").hide();
		}

		frappe.show_alert({
			message: __("Auto-filled {0} field(s)", [filled]),
			indicator: filled ? "green" : "orange",
		});
	}

	page.body.find("#extractBtn").on("click", function () {
		const front = frontCtrl.get_value();
		const back = backCtrl.get_value();
		if (!front && !back) {
			frappe.msgprint(__("Please upload the front or back of the visiting card."));
			return;
		}

		if (!aiSettings.enable_openai || !aiSettings.openai_configured) {
			frappe.msgprint({
				title: __("AI not configured"),
				message: __("Enable OpenAI and set API key + model in Pratap CRM Ai Settings."),
				indicator: "orange",
			});
			return;
		}

		const $btn = $(this);
		resetForm(false, true);
		page.body.find("#ocr-status").text(__("Extracting contact details using AI…"));
		$btn.prop("disabled", true);

		frappe.call({
			method: `${PAGE_METHOD}.extract_business_card_data`,
			args: { front_image: front || null, back_image: back || null },
			freeze: true,
			freeze_message: __("Analyzing business card…"),
			callback(r) {
				$btn.prop("disabled", false);
				const msg = r.message;
				if (!msg || msg.status === "disabled") {
					page.body
						.find("#ocr-status")
						.html(`<span class="text-danger">${__("AI extraction is not available.")}</span>`);
					return;
				}

				page.body.find("#ocr-raw-text").text(msg.raw_text || "");
				page.body.find("#ocr-raw-card").show();
				fillProspectFields(msg.parsed_data || {});
				page.body.find("#ocr-status").text(__("Data extracted successfully."));
			},
			error(err) {
				$btn.prop("disabled", false);
				page.body
					.find("#ocr-status")
					.html(
						`<span class="text-danger">${frappe.utils.escape_html(err.message || "")}</span>`
					);
			},
		});
	});

	page.body.find("#toggle-ocr").on("click", function () {
		const $raw = page.body.find("#ocr-raw-text");
		$raw.toggle();
		$(this).text($raw.is(":visible") ? __("Hide") : __("Show"));
	});

	page.body.find("#saveProspectBtn").on("click", () => {
		const company_name = (prospectControls.company_name.get_value() || "").trim();
		if (!company_name) {
			frappe.msgprint(__("Company Name is required."));
			return;
		}

		const prospect_data = {};
		for (const key of PROSPECT_FIELD_KEYS) {
			prospect_data[key] = (prospectControls[key].get_value() || "").trim();
		}

		saveProspect(prospect_data);
	});

	function saveProspect(prospect_data) {
		frappe.call({
			method: `${PAGE_METHOD}.create_prospect`,
			args: {
				prospect_data: JSON.stringify(prospect_data),
				front_image: frontCtrl.get_value() || null,
				back_image: backCtrl.get_value() || null,
			},
			freeze: true,
			freeze_message: __("Creating Prospect…"),
			callback(r) {
				const $msg = page.body.find("#prospect-insert-msg");
				$msg.empty();
				if (r.message) {
					showProspectSuccessBanner(r.message);
					resetForm(false, false, false);
					frappe.show_alert({
						message: __("Prospect saved successfully"),
						indicator: "green",
					});
				} else {
					page.body.find("#prospect-success-slot").empty();
					$msg.html(
						`<div class="alert alert-danger">${__("Failed to create Prospect.")}</div>`
					);
				}
			},
			error(err) {
				if (err?.message) {
					page.body.find("#prospect-insert-msg").html(
						`<div class="alert alert-danger">${frappe.utils.escape_html(err.message)}</div>`
					);
				}
			},
		});
	}
};
